### DISCUSSION TITLE
OpenApi 3.0 spec for API?

### DISCUSSION THREAD

**Discussion by giautm** [pocketbase/pocketbase#209] [Q&A]
I want to access the APIs via Golang/PHP. If we can provide the specification via OpenAPI spec. I can generate the client using existing tools. You also generate the server code using OpenAPI tools.
For example: https://github.com/ogen-go/ogen

**Comment by ganigeorgiev** [comment]
Although I can see the benefit of OpenApi (in the past I've even written a php generator for the v2 spec), in reality the end result may not be very practical and it could also sometimes obfuscate the code if you keep the documentation as comments alongside the handlers.

I don't think generating an OpenApi docs for PocketBase will be very useful in general because developers most likely are going to use mainly the [Records](https://pocketbase.io/docs/api-records/) and the [Realtime](https://pocketbase.io/docs/api-realtime/) services and they are fairly small and easy to wrap. Additionally, since the collection fields are user defined, we'll have to keep the generated OpenApi schema somehow in sync during runtime (as a hook with cache?), which may complicate further the internals.

In short - I'm against adding support for OpenApi for PocketBase since it will be hard to maintain and keep in sync with the existing more human readable documentation and examples (which will be expanded even further in the near future).

I'll convert the issue to discussion and if someone disagree and would like to see OpenApi supported, please feel free to share your feedback (it really helps me to understand better how developers prefer to work with PocketBase).

**Reply by giautm** [reply]
Hi, thanks for your kindly response.

I'm new to PocketBase, I wanna try this project for my projects written in Go. So, I have to write the client to connect to PocketBase. The base OpenAPI spec can have some basic fields from your design (example: pagination, filter. Collection creates). For any user-defined field, let them put it in the spec went they need.

```json
{
  "page": 1,
  "perPage": 100,
  "totalItems": 2,
  "items": [
    {
      "@collectionId": "a98f514e-b05f-4544-a93d-8981c7990728",
      "@collectionName": "demo",
      "id": "ae40239d-2bc4-4775-8613-4c3693d585b0",
      "updated": "2022-06-25 11:03:50.052",
      "created": "2022-06-25 11:03:35.163",
      "title": "test1" // <--- this field should added by the user
    },
    {
      "@collectionId": "a98f514e-b05f-4544-a93d-8981c7990728",
      "@collectionName": "demo",
      "id": "088dfc4f-4d84-4193-a4e5-1f6716c3726f",
      "updated": "2022-06-25 11:03:45.876",
      "created": "2022-06-25 11:03:45.876",
      "title": "test2" // <--- this field should added by the user
    }
  ]
}
```

**Reply by giautm** [reply]
Or, we can provide an endpoint on PocketBase that generate OpenAPI spec on the air when the user request. It will allow the client always be consistent with the schema/Collection.

In some use cases (I think), the collection doesn't change too much after it is created.

**Reply by ganigeorgiev** [reply]
Yes, that's another way to have it generated.

It is doable but again it is one extra thing that we will have to care for and I'm not sure whether it is worth investing the time since the generated client will be very basic (I doubt that it would differ too much from the std http client in Go) and probably you'll have to modify/adjust it anyway (especially when handling file uploads or the realtime sse subscriptions).

Furthermore, for Go you could import PocketBase directly in your existing application and interact with its internals without having to go through the network (there isn't much documentation and examples for this at the moment, but it'll be added in the near future).

To be honest, I'm more eager to consider building a proper SDK for the specific language since it'll cover more use cases, but that depends on the users demand for it.

**Reply by pritamsoni-hsr** [reply]
I feel like this is a very powerful and useful feature for most people to iterate their apps, and be aware of breaking changes before pushing things out.

I'm starting [this](http://github.com/pritamsoni-hsr/pocketbase-codegen) today. I'll need to check how [swagger parser](https://github.com/go-swagger/go-swagger) parses types interally, I'm not very much familiar with it tbh, or [goa](https://pkg.go.dev/goa.design/goa/v3/pkg)

Doing this will also help build a better sdk, for instance you can use a local or remote schema to get hints about all available collections, and so on, this will improve developer experience for those using sdk.

```ts
client.records.getList({collection:'<name>', page:1, pageSize:10})
```

**Reply by pritamsoni-hsr** [reply]
I got it to be able to generate the openapi schema, which can be used to generate clients, and an accessible SDK. 
There is some extra garbage that needs to be removed.

<details>
  <summary>Schema here</summary>

```json
{
  "swagger": "2.0",
  "info": { "title": "PocketBase API", "version": "0.0.1" },
  "host": "localhost:8090",
  "consumes": ["application/json", "application/xml", "application/gob"],
  "produces": ["application/json", "application/xml", "application/gob"],
  "paths": {
    "/api/collections/profiles/records": {
      "get": {
        "tags": ["profiles"],
        "summary": "list profiles",
        "operationId": "profiles#list",
        "responses": {
          "200": { "description": "OK response.", "schema": { "type": "array", "items": { "$ref": "#/definitions/profilesResponse" } } }
        },
        "schemes": ["http"]
      },
      "post": {
        "tags": ["profiles"],
        "summary": "create profiles",
        "operationId": "profiles#create",
        "parameters": [
          {
            "name": "CreateRequestBody",
            "in": "body",
            "required": true,
            "schema": { "$ref": "#/definitions/ProfilesCreateRequestBody", "required": ["userId"] }
          }
        ],
        "responses": {
          "200": {
            "description": "OK response.",
            "schema": { "$ref": "#/definitions/ProfilesCreateResponseBody", "required": ["userId"] }
          }
        },
        "schemes": ["http"]
      }
    },
    "/api/collections/profiles/records/:id": {
      "get": {
        "tags": ["profiles"],
        "summary": "view profiles",
        "operationId": "profiles#view",
        "responses": {
          "200": { "description": "OK response.", "schema": { "$ref": "#/definitions/ProfilesViewResponseBody", "required": ["userId"] } }
        },
        "schemes": ["http"]
      },
      "delete": {
        "tags": ["profiles"],
        "summary": "delete profiles",
        "operationId": "profiles#delete",
        "responses": { "204": { "description": "No Content response." } },
        "schemes": ["http"]
      },
      "patch": {
        "tags": ["profiles"],
        "summary": "update profiles",
        "operationId": "profiles#update",
        "parameters": [
          {
            "name": "UpdateRequestBody",
            "in": "body",
            "required": true,
            "schema": { "$ref": "#/definitions/ProfilesUpdateRequestBody", "required": ["userId"] }
          }
        ],
        "responses": {
          "200": {
            "description": "OK response.",
            "schema": { "$ref": "#/definitions/ProfilesUpdateResponseBody", "required": ["userId"] }
          }
        },
        "schemes": ["http"]
      }
    },
    "/openapi": {
      "get": {
        "tags": ["profiles"],
        "summary": "Download profiles.json",
        "operationId": "profiles#/openapi",
        "responses": { "200": { "description": "File downloaded", "schema": { "type": "file" } } },
        "schemes": ["http"]
      }
    }
  },
  "definitions": {
    "ProfilesCreateRequestBody": {
      "title": "ProfilesCreateRequestBody",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "Maxime consectetur." },
        "name": { "type": "string", "example": "Autem eum sed voluptas vel distinctio commodi." },
        "userId": { "type": "string", "example": "Qui voluptatibus veritatis." }
      },
      "example": {
        "avatar": "Eligendi velit vitae deleniti quia exercitationem accusantium.",
        "name": "Asperiores dolore maxime non pariatur.",
        "userId": "Blanditiis qui fuga est accusamus autem."
      },
      "required": ["userId"]
    },
    "ProfilesCreateResponseBody": {
      "title": "ProfilesCreateResponseBody",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "Et nisi hic esse doloribus deserunt id." },
        "name": { "type": "string", "example": "Reiciendis quas consequatur rerum facere." },
        "userId": { "type": "string", "example": "Aut praesentium iusto rerum odit." }
      },
      "example": { "avatar": "Quia non aspernatur.", "name": "Sed quas ex.", "userId": "Deserunt est voluptas ut." },
      "required": ["userId"]
    },
    "ProfilesUpdateRequestBody": {
      "title": "ProfilesUpdateRequestBody",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "Reiciendis ipsam repellat veritatis." },
        "name": { "type": "string", "example": "Est sunt." },
        "userId": { "type": "string", "example": "Minus optio quisquam." }
      },
      "example": {
        "avatar": "Beatae magni incidunt molestias consectetur.",
        "name": "Architecto eligendi laborum consequatur qui omnis.",
        "userId": "Qui sed."
      },
      "required": ["userId"]
    },
    "ProfilesUpdateResponseBody": {
      "title": "ProfilesUpdateResponseBody",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "Nam magni." },
        "name": { "type": "string", "example": "Est fugit." },
        "userId": { "type": "string", "example": "Sequi sunt ut recusandae perspiciatis ea eos." }
      },
      "example": { "avatar": "Ipsam accusamus nemo molestiae architecto beatae.", "name": "Et nulla id.", "userId": "Eligendi ducimus." },
      "required": ["userId"]
    },
    "ProfilesViewResponseBody": {
      "title": "ProfilesViewResponseBody",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "Optio consequatur minima." },
        "name": { "type": "string", "example": "Dolores expedita accusantium iste in facilis ea." },
        "userId": { "type": "string", "example": "Nam officia ut et." }
      },
      "example": {
        "avatar": "Doloribus aut et assumenda et accusamus temporibus.",
        "name": "Occaecati eveniet et architecto praesentium.",
        "userId": "Nisi est consequatur."
      },
      "required": ["userId"]
    },
    "profilesResponse": {
      "title": "profilesResponse",
      "type": "object",
      "properties": {
        "avatar": { "type": "string", "example": "A quidem." },
        "name": { "type": "string", "example": "Eos non deserunt aut velit sunt." },
        "userId": { "type": "string", "example": "Quam quisquam maiores libero." }
      },
      "example": {
        "avatar": "Voluptas assumenda occaecati in aut.",
        "name": "Est sunt aut cumque repudiandae.",
        "userId": "Eligendi veritatis sunt."
      },
      "required": ["userId"]
    }
  }
}
```
</details>

**Reply by regisin** [reply]
I'd love to have an openapi spec so I can write the python client myself. I know there are python sdks out there, but my use case is different. Having a swagger page with all the available routes would really help me. For example, I have no idea what the route for client auth is... the only docs I see are the dart and js sdks.

**Reply by ganigeorgiev** [reply]
@regisin The Web APIs documentation could be found under the "Web APIs" section - https://pocketbase.io/docs/api-records/.

Alternatively, you can also find documentation for each individual collection from the Admin UI > API Preview:
![preview](https://i.imgur.com/NeBxF1g.png)

**Reply by regisin** [reply]
I see! Yes, it's there. Although I'd sugest changing `/api/collections/collectionIdOrName/auth-with-password` to `/api/collections/users/auth-with-password`? I looked at it and saw a generic `collectionIdOrName`. It's at https://pocketbase.io/docs/api-records/ under Auth with password.

**Reply by ganigeorgiev** [reply]
With PocketBase v0.8.0 we now support multiple auth collections, aka. you can have users, members, staffs, etc., each with their own custom fields and different login options.

The above screenshot is the API preview from the `users` collection and this is why it has `users` in the url. The general form is `collectionIdOrName`.

**Reply by regisin** [reply]
Ok, another overlook on my end. I had v0.7, didn't realize 0.8 was released. All good now, thanks.

**Reply by dschissler** [reply]
> With PocketBase v0.8.0 we now support multiple auth collections, aka. you can have users, members, staffs, etc., each with their own custom fields and different login options.
> 

@ganigeorgiev How would this work for signin when each type of user is stored in their own collection?  Would they need to each login via their own special page according to their group / collection membership?  Or would you first in code have them try to sign into "staff" and then if that failed try "members" and then finally "users" before considering it a login failure?

Also what happens if their login status changes?  So then would you have to copy their details over to another collection?  Unless I'm missing something, this seems like a bad idea compared to using a relation from user to group collection.

So I understand that you're just trying to offer flexibility to developers who want to use Pocketbase for gluing many application spaces together, but I'm struggling to see how your example makes any sense.  Am I missing a pattern or concept for tying all of these auth collections together?

**Reply by ganigeorgiev** [reply]
@dschissler 

> How would this work for signin when each type of user is stored in their own collection? Would they need to each login via their own special page according to their group / collection membership? Or would you first in code have them try to sign into "staff" and then if that failed try "members" and then finally "users" before considering it a login failure?

It's up to the developers how they will handle it. You can have individual login pages (eg. `/users/login` and `/staffs/login`) or you can use the failover flow (aka. try first login with the `staffs` collection, if fail - try `users`).

---

> Also what happens if their login status changes? So then would you have to copy their details over to another collection? Unless I'm missing something, this seems like a bad idea compared to using a relation from user to group collection.

I'm not sure that I understand this. The auth collections are not "roles". **If your app users have the same set of fields and only a single role/group/team field is different, then you don't need multiple auth collections.**

---

> So I understand that you're just trying to offer flexibility to developers who want to use Pocketbase for gluing many application spaces together, but I'm struggling to see how your example makes any sense. Am I missing a pattern or concept for tying all of these auth collections together?

Personally I added support for multiple auth collections because I needed it at work for a small internal wiki where staff and clients has their own login pages and both user types have completely different fields with different constraints.
